levels/fullstack.tex

```diff
@@ -0,0 +1,322 @@
+    data structures in day to day activity.
+
+    \bigbreak
+
+    Can implement a production quality software - it might be not the most
+    efficient or secure, but correct.
+    &
+
+    % Security
+    Understands basics of web security (can configure TLS on a standalone web
+    server like nginx).
+    &
+
+    % Networking
+    Understands and reasons about browser networking concepts. Understands
+    common networking issues and can troubleshoot issues using the developer
+    console.
+    &
+
+    % Fullstack Engineering
+    Understands Javascript, HTML, CSS, and DOM. Can fix bugs or add small
+    features to frontend applications.
+    &
+
+    % Tooling
+    Understands and can use build tools (like npm and yarn), Docker, and
+    Makefiles.
+    \\ [13em]
+  \hline
+    % Level
+    2
+    &
+
+    % Product Output
+    \cellcolor{blue!25}
+    Can write high quality user and product focused documentation.
+    &
+
+    % Communication/Writing
+    Provides constructive review on peers' code and design. Helps new team
+    members during their first weeks.
+    &
+
+    % Data structures/Performance
+    &
+
+    % Security
+    Understands how to utilize browser security features (like headers and
+    cookie attributes) to secure web applications.
+    &
+
+    % Networking
+    Has more granular understanding of the networking design, for example can
+    reason about using gRPC vs HTTP/JSON and websockets and their networking
+    and scalability trade-offs.
+    &
+
+    % Fullstack Engineering
+    Can write frontend applications applications.
+    &
+
+    % Tooling
+    Understands and can use transpilers (TypeScript to Javascript) and SPA
+    frameworks (React, Angular). Can leverage components and styles from
+    third-party libraries to implement designs.
+    \\ [13em]
+  \hline
+    % Level
+    3
+    &
+
+    % Product Output
+    \cellcolor{blue!25}
+    Collaborates with the team to scope requirements, based on good
+    understanding of existing longer term product vision and estimates of the
+    system design of a feature of a product.
+    &
+
+    % Communication/Writing
+    Supports less experienced peers' technical skills, answering questions and
+    being a resource. Documents and improves team practices.
+    &
+
+    % Data structures/Performance
+    Understands browser-based event handling and event propagation.
+    Understands the use and implications of closures.
+    &
+
+    % Security
+    Can use frameworks to mitigate common attack vectors for server side (like
+    secure password storage) or client side web applications (like XSS, CSRF, and
+    SSRF) security vulnerabilities.
+    &
+
+    % Networking
+    Understands scalability and resilience aspects of practical implementation
+    of systems leveraging networking.
+
+    \bigbreak
+
+    Understands certificate-based mutual authentication network protocols like
+    mTLS and SSH.
```

Comment:
It seems that fullstack.pdf needs to be updated as this change doesn't show up in the file.